### ISSUE TITLE
XQuery 3

### ISSUE THREAD

**Issue by ajnewton1** [ebeshero/DHClass-Hub#290]
I dont understand what we are supposed to do for question 3. Are we supposed to do it in oXygen or in eXide?

**Comment by ebeshero** [comment]
Hi @ajnewton1 ....Question 3 is "Download the KML file from the GPS Visualizer site, and save it to work with it. We will now make some changes to it so that it includes more information that we will extract via XQuery from the Church Schism project files. For this assignment, we would like to capture the text of the longest paragraph surrounding each location to hold in the placemarks of our KML, so that this text appears in a pop-up window when we click the location pins on our map." 

Basically this is just describing what we want to do with the file you downloaded. We want to add some information to it. 

Your next step is to go and make the the other KML file that holds the placenames + the paragraph that holds them in the project files. You'll need to make that in eXide, and when you've output that, you'll go to work in oXygen to braid the two files together with that XSLT identity transform!

**Comment by Samantha-Mcguigan** [comment]
I'm trying to braid the two files together but my output is not right. It is outputting a self-closing `</description>` tag and I don't know why. I input the KML file we made yesterday using the GPS visualizer and used the KML-IDTransform.xsl file. I also made sure to change the Placeography.kml file to have `<kml xmlns="http://www.opengis.net/kml/2.2">` at the top instead of just `<kml>` and they are all in the same folder so I don't understand why it is not working.

**Comment by ajnewton1** [comment]
I am having difficulties with what to put in the `</description>` tag. This is what I have so far... Is it correct? 
`<kml>
{
let $schism := collection('/db/Schism')/*
let $loc := $schism//location
let $distLoc := distinct-values($loc)
for $i in $distLoc
return
    <Placemark>
        <name>{$i, '&#10;'}</name>
        <description>{$loc/ancestor::p[contains(., $distLoc)][string-length() = max($loc/ancestor::p[contains(., $distLoc)]/string-length())]}
        </description>
    </Placemark>
}
</kml>`

**Comment by ebeshero** [comment]
@Samantha-Mcguigan Hmmm. It sounds like your namespace on the Placeography file is right, so I think the problem might be something with the template match. Check for something small like, is the description element spelled the same in the XSLT template match and in your file? Is it sitting where it's supposed to be? 

Try pushing your whole Placeography file in the Troubleshooting directory and we can help see if we can spot the problem...Also, did you download the XSLT from here on DHClass-Hub?

**Comment by ebeshero** [comment]
@ajnewton1 I'm studying your code...just put four tick marks above and below to see if better...

**Comment by ajnewton1** [comment]
````
<kml>
{
let $schism := collection('/db/Schism')/*
let $loc := $schism//location
let $distLoc := distinct-values($loc)
for $i in $distLoc
return
    <Placemark>
        <name>{$i, '&#10;'}</name>
        <description>{$loc/ancestor::p[contains(., $distLoc)][string-length() = max($loc/ancestor::p[contains(., $distLoc)]/string-length())]}
        </description>
    </Placemark>
}
</kml>````

**Comment by ajnewton1** [comment]
Wait I think I got it but I am not sure.
````
<kml>
{
let $schism := collection('/db/Schism')/*
let $loc := $schism//location
let $distLoc := distinct-values($loc)
for $i in $distLoc
let $max := $loc/ancestor::p[contains(., $i)][string-length() = max($loc/ancestor::p[contains(., $i)]/string-length())]
return
    <Placemark>
        <name>{$i, '&#10;'}</name>
        <description>{$max}</description>
    </Placemark>
}
</kml>
````

I think this gives the right output @ebeshero

**Comment by ebeshero** [comment]
@ajnewton1 I was just typing that I think you mean to have $i and not $distloc! And you figured it out. Good!

**Comment by ebeshero** [comment]
@ajnewton1 It's interesting you used the contains() function for this--That would get you the longest paragraph that contains the string of $i, but might actually have more in it--like for $i=Pitt you might get a paragraph containing Pittsburgh! 

It might work out fine, or it might be what we'd call a fuzzy match! Instead you could filter based on equivalence...

**Comment by ajnewton1** [comment]
Now for some reason it is not letting me save it as a file. It is giving me an error on ````let $dob-db-uri := xmldb:store("/db/ajnewton1/", $filename, $thisFile)```` saying that I don't have permission on the collection.

**Comment by RJP43** [comment]
@ajnewton1 Make sure are logged in. Even if it says you are you might log out and log back in ... it can be wonky sometimes.

**Comment by ebeshero** [comment]
@ajnewton1 see my comment just ahead of yours re contains()...
Are you logged in to eXist? Sounds like you might not be.

**Comment by ajnewton1** [comment]
Actually, you might be right @ebeshero. When I look through all my data it looks like my Allegheny County text doesn't match up with your example

**Comment by ebeshero** [comment]
@ajnewton1 Well, my example is outdated--from an older version of this assignment where I had everyone grab the **first** paragraph holding the location, not the **longest**.  So you might be fine, but you might save this version AND try processing a different way to compare the results!

**Comment by Samantha-Mcguigan** [comment]
So there were some extra spaces inside my element tags before the closing tag and I deleted them so now the entire description is showing up but when I open my new output KML file I am getting 45 errors on all the opening `<name>` elements?

**Comment by ebeshero** [comment]
@Samantha-Mcguigan Uh oh! That's an interesting problem! What is the error message?

**Comment by Samantha-Mcguigan** [comment]
It says invalid content was found starting with element '{"http://www.opengis.net/kml/2.2":name}

**Comment by ebeshero** [comment]
Hmmm...Sounds like something still isn't processing right...Can you push your output file to the Troubleshooting folder here on DH-ClassHub?

**Comment by Samantha-Mcguigan** [comment]
yes just a second

**Comment by ebeshero** [comment]
@Samantha-Mcguigan I have a hunch that the elements might be wrapped problematically in your Placeography file... ?

**Comment by Samantha-Mcguigan** [comment]
its in!

**Comment by ebeshero** [comment]
Taking a look...In the meantime, see comments above...

**Comment by Samantha-Mcguigan** [comment]
I also used contains( ) so maybe that is a problem?

**Comment by ebeshero** [comment]
@Samantha-Mcguigan Probably not...It will be interesting to compare the results.  Your Placeography file looks okay, but the first thing I notice in your output is that the `<description>` element comes out first, BEFORE the `<name>` element. That's what is causing the error. Now, I'm not sure why that's happening, so I'm going to run my XSLT on your files and see if I can figure out why...

**Comment by Samantha-Mcguigan** [comment]
this is what I have in XQuery if it helps:

```
xquery version "3.0";
<kml>
    {
let $Schism := collection('/db/Schism')/*
let $locs := $Schism//location
let $distLocs := distinct-values($locs)
for $i in $distLocs
return
    <Placemark>
    <name>{$i, '&#10;'}</name>
<description>{max(//p[contains(., $i)]/string())}</description>
</Placemark>
}
 </kml>
```

**Comment by ajnewton1** [comment]
What is the XML file supposed to be when we try to output using the identity transformation as the XSLT

**Comment by ebeshero** [comment]
@ajnewton1 You can name that whatever you like--it's what you'll be opening up in Google Maps. 
@Samantha-Mcguigan I think your output looks right, and I'm realizing that my XSLT is producing the code just as I wrote it, which means that when we template match on the name in the original file, we replace it with the description, so the description appears first.

And when we template match on the description in the original file, we *rename* it as `<name>` because the output description from the GPS visualizer gave us the name.

I think what you've discovered is that we're producing technically invalid KML as our output...I bet it still works in Google Maps, but I should debug this. @RJP43 Want to check your output KML and verify--do you see the same error as @Samantha-Mcguigan ? I'll check my files, too.

**Comment by ajnewton1** [comment]
<img width="703" alt="screen shot 2017-03-14 at 11 37 52 pm" src="https://cloud.githubusercontent.com/assets/24962219/23932570/4915b914-090f-11e7-83fe-48a24c6e93d7.png">
Is this how we are supposed to output the KML file from the identity transformation?

**Comment by ebeshero** [comment]
@ajnewton1 Not quite...You're going to transform the KML produced by the GSPS Visualizer! But the Placeography file you made is being woven in--we're calling it in the Identity Transformation as a variable using the `document()` function.

**Comment by Samantha-Mcguigan** [comment]
It is working on google maps, I'm going to upload what I have to courseweb and got to bed :)

**Comment by ajnewton1** [comment]
I am getting an error with ````<name>Nevada</name>```` in the map.kml and ````<xsl:value-of select="$PO//name[. = current()]/following-sibling::description"/>```` in the xslt sheet

**Comment by ebeshero** [comment]
@ajnewton1 What's the error? Check to see that you put the namespace in your Placeography.kml file!

**Comment by ebeshero** [comment]
@Samantha-Mcguigan That's fine! I've got a patch we can run in class tomorrow: It's basically just one more XSLT Identity Transformation that you'd run on your output to put the `<name>` and the `<description>` element in the right order! I'm testing it now.

**Comment by ajnewton1** [comment]
Its an error reported by XML parser processing my file and the namespace is in that file

**Comment by ajnewton1** [comment]
err:FODT0002, overflow in duration arithmetic

**Comment by ajnewton1** [comment]
I figured it out. In the XSLT sheet it was importing 'Placeography.kml' but my file for some reason automatically saved as 'Placeography.kml.xml'

**Comment by ebeshero** [comment]
@ajnewton1 I bet your output is going to have the same errors as Sam's...if it does, I just prepared a patch XSLT you can run. (I'm pushing it to Class Examples in a second.) Let me know how your output looks (save it and reopen in oXygen.

**Comment by ajnewton1** [comment]
Yes you are right. It is having errors on all the ``<name>`` tags

**Comment by ebeshero** [comment]
@ajnewton1 Gimme a second, and I'll push the Patch file you can run on it! :-)

**Comment by ebeshero** [comment]
@ajnewton1 I just pushed my patch file: It's in the Class Examples -> KML directory. You can pull it in, save it in your working directory, and run it on your output file with the 45 errors on the name tags!

**Comment by Samantha-Mcguigan** [comment]
Courseweb won't let me upload my .txt files for some reason, it says they are a security threat (*rolls eyes) can I email them to you or push them to GitHub?

**Comment by ajnewton1** [comment]
I ran this and ended up with some strange errors 
<img width="412" alt="screen shot 2017-03-15 at 12 12 36 am" src="https://cloud.githubusercontent.com/assets/24962219/23933338/3fb631d2-0914-11e7-93e6-a28c69489dc4.png">
This is happening to all the ``<Placemark>`` elements

**Comment by ebeshero** [comment]
@ajnewton1 Check your source file for the transformation: which KML file do you have selected for your input? It should be the output KML with the 45 errors on name...

**Comment by ajnewton1** [comment]
Ahhh makes sense! It all looks shiny and clean now. Thanks @ebeshero !

**Comment by ebeshero** [comment]
@Samantha-Mcguigan Sure--that's fine. Or put them in Box where your homework is returned!

**Comment by Samantha-Mcguigan** [comment]
They are in Box!!

**Comment by ebeshero** [comment]
Thanks @Samantha-Mcguigan 
Everyone--if you're having trouble uploading a homework file on Courseweb, just upload them to Box!

**Comment by ajnewton1** [comment]
That happened to me as well but I put them in a zipped folder and uploaded to course web

**Comment by ebeshero** [comment]
That should work okay, too! @ajnewton1

**Comment by ahunker** [comment]
@ebeshero 
 We de-bugged the XSLT today after class. Note for the assignment sheet, when outputting the Placeography.kml, wrap all Placemark elements in a Folder element to output valid kml

**Comment by jonhoranic** [comment]
@ebeshero For some reason both the old transformation file and the fixed transformation file will not move the descriptions over correctly and it is baffling, I am unsure of how to proceed. The map works though! I tested that and each time it was fine, just no descriptions!

**Comment by ebeshero** [comment]
@jonhoranic Did you apply the namespace to your Placeography.kml before running the XSLT identity transformation?

**Comment by jonhoranic** [comment]
@ebeshero Sorry for the late response, for some reason GitHub did not ping or update to show me you posted. To answer that, yes the namespace was present and I kept trying different things to see if it would help. I turned it in via courseweb if you wish to see what has gone wrong.
